resolwe/storage/manager.py

```diff
@@ -0,0 +1,188 @@
+"""Storage manager"""
+from datetime import datetime, timedelta
+import logging
+from typing import List, Tuple
+
+from django.conf import settings
+from django.db import transaction
+from django.utils.timezone import now
+
+from resolwe.storage.models import FileStorage, StorageLocation, AccessLog
+from resolwe.storage.connectors import connectors, Transfer
+
+
+logger = logging.getLogger(__name__)
+
+
+class DecisionMaker:
+    """Decide what to do with the FileStorage object.
+    Consider all knows connectors and rules for moving between them.
+    """
+
+    def __init__(self, file_storage: FileStorage):
+        self.file_storage = file_storage
+
+    def _get_rules(self, connector_name: str, _type: str) -> dict:
+        """Get rules for the given location. Valid types of rules are "copy"
+        and "delete".
+        """
+        location_settings = settings.STORAGE_CONNECTORS.get(connector_name, {})
+        return location_settings.get("config", {}).get(_type, {})
+
+    def copy(self) -> List[str]:
+        """Decide if the file storage object should be copied to more storage
+        locations.
+
+        :return: a list connector names where data should be copied to.
+        :rtype: list[str]
+        """
+        return [
+            connector_name
+            for connector_name in settings.STORAGE_CONNECTORS
+            if self._should_copy_to(connector_name)
+        ]
+
+    def _should_copy_to(self, connector_name: str) -> bool:
+        """Decide if data should be copied to the given storage connector
+        accoring to the rules.
+
+        :param connector_name: name of the storage connector.
+        :type connector_name: str
+
+        :return: True when data should be copied to the location managed by
+            connector connector_name.
+        """
+        copy_rules = self._get_rules(connector_name, "copy")
+
+        rule_results = {}
+
+        rule_results["has_rules"] = len(copy_rules) > 0
```

Comment:
When there are no copy rules defined I believe that we should not copy data to that connector. 

Since all returns True for empty list, I added one to make sure it will not pass with empty rule set.

Reply:
I don't envision any use-case where a connector is defined but should not be used. Correct me if I am wrong.

My understanding was that data are immediately copied to the connector unless there is an extra copy rule to postpone it.

Reply:
I vote for explicit copy rule. Implicit "always copy" rule could cause troubles.

When there is no copy rule for the "local" connector (and I believe there should be none, since workers will pull data when needed) then everything located other connectors and not on the local one will be immediately copied to the local connector.